src/com/palmergames/bukkit/TownyChat/listener/TownyChatPlayerListener.java

```diff
@@ -116,22 +116,27 @@ public void onPlayerChat(AsyncPlayerChatEvent event) {
 			 */
 			if (directedChat.containsKey(player)) {
 				Channel channel = plugin.getChannelsHandler().getChannel(player, directedChat.get(player));
-				directedChat.remove(player);
 				
 				if (channel != null) {
 					// Notify player he is muted
 					if (channel.isMuted(player.getName())) {
 						TownyMessaging.sendErrorMsg(player, String.format(Translation.of("tc_err_you_are_currently_muted_in_channel"), channel.getName()));
 						event.setCancelled(true);
+						directedChat.remove(player);
 						return;
 					}
 					if (channel.isSpam(player)) {
 						event.setCancelled(true);
+						directedChat.remove(player);
 						return;
 					}
 					channel.chatProcess(event);
+					if (!Chat.usingEssentialsDiscord || event.isCancelled()) {
+						directedChat.remove(player);
+					}
 					return;
 				}
+				directedChat.remove(player);

```

Comment:
Not entirely sure why this change was made, directedChat.remove(player) should always occur, as this is when a player uses `/tc somechat` in order to speak into a channel which isn't the one they're currently speaking in.

Reply:
This is done so `DiscordChatMessageEvent` can check to see if it was a direct channel, it will always end up being removed though, it's just done in `DiscordChatMessageEvent` if EssentialsDiscord was found.